### DISCUSSION TITLE
Feature Request:  Build wpa_supplicant with CONFIG_80211R=y

### DISCUSSION THREAD

**Discussion by dwrice0** [WLAN-Pi/feedback#9] [General Inquiries] [answered]
Not sure if you're building wpa_supplicant from source or just installing the package direct from a repo, but if you are building it, I have a very small request for the next release:  can you change .config to add CONFIG_80211R=y?  Latest wlanpi 2.0 image has wpa_supplicant 2.8 which can support FT but it wouldn't join an FT PSK network so likely wasn't built with that flag.  I built wpa_supplicant 2.9 from source with that flag and it joins FT-PSK now.  Not that wlanpi really needs FT as I don't know anyone using it for real-time apps.  I was using it to join FT-PSK so I could capture the keying process of an FT 4-way handshake (over the air AND debugs with the wpa_supplicant -dd -K options).

**Comment by jolla** [comment, marked as answer]
It looks like CONFIG_IEEE80211R=y is the default in wpa_supplicant v2.9. Does this look correct?

```
# IEEE Std 802.11r-2008 (Fast BSS Transition) for station mode
CONFIG_IEEE80211R=y
```

**Reply by dwrice0** [reply]
To be honest, I didn't ever build 2.9 w/o the flag in .config

I actually started with wlanpi 1.x and it has wpa_supplicant 2.4 and in my research found that 802.11r wasn't supported until 2.6. So, I upgraded to wlanpi 2.x to see what version was included there and it has 2.8 by default.

When at 2.8 it wouldn't join an FT PSK, I googled some more and ran across a few posts that said the flag was needed at build time even though it was "supported".  So, I just grabbed the latest stable (2.9) and built it with the flag to begin with.

**Reply by jolla** [reply]
Ok, currently we aren't compiling wpa_supplicant from source so the easiest solution would be to upgrade wpa_supplicant to v2.9 if that would add this by default. Could you confirm and report back if that is the case?

**Reply by dwrice0** [reply]
Will do.

**Reply by dwrice0** [reply]
OK.  wpa_supplicant 2.9 package included in debian bullseye repo will not join an 802.11r only BSS.  In fact, it errors out reading the wpa_supplicant.conf file if you have a network with key_mgmt=FT-PSK with an error message regarding unknown key management type.  If you build from source with CONFIG_IEEE80211R=y w/o changing wpa_supplicant.conf, the same hardware will join the 802.11r-only BSS.

**Reply by jolla** [reply]
Ok, thanks for testing and confirming

**Reply by danielmundi** [reply]
@dwrice0, I was just checking Debian bullseye wpa_supplicant build configuration and it seems they are adding CONFIG_IEEE8022R into the config. I just want to make sure there isn't anything else missing from the build in Debian repo.

Can you share the results of the 4 commands below using wpa_supplicant installed from Debian?
```bash
$ strings $(which wpa_supplicant) | grep FT-PSK
$ dpkg -s wpasupplicant
$ cat /etc/apt/sources.list
$ cat /etc/apt/sources.list.d/*
```

Thanks!

**Reply by dwrice0** [reply]
```
wlanpi@wlanpi:~/source$ strings $(which wpa_supplicant) | grep FT-PSK
%sFT-PSK
FT-PSK
 FT-PSK
wlanpi@wlanpi:~/source$ dpkg -s wpasupplicant
Package: wpasupplicant
Status: install ok installed
Priority: optional
Section: net
Installed-Size: 3343
Maintainer: Debian wpasupplicant Maintainers <wpa@packages.debian.org>
Architecture: arm64
Multi-Arch: foreign
Source: wpa
Version: 2:2.9.0-21
Depends: libc6 (>= 2.29), libdbus-1-3 (>= 1.9.14), libnl-3-200 (>= 3.2.21), libnl-genl-3-200 (>= 3.2.7), libnl-route-3-200 (>= 3.4.0), libpcsclite1 (>= 1.0.0), libreadline8 (>= 6.0), libssl1.1 (>= 1.1.1), lsb-base, adduser
Suggests: wpagui, libengine-pkcs11-openssl
Breaks: initscripts (<< 2.88dsf-13.3)
Conffiles:
 /etc/dbus-1/system.d/wpa_supplicant.conf 3d46000a332322a8c1b432df148a2d6b
 /etc/wpa_supplicant/action_wpa.sh 41b2e273d616b1c7e7b605bbe8648d8b
 /etc/wpa_supplicant/functions.sh 29834d0d67a65585b71e231832884174
 /etc/wpa_supplicant/ifupdown.sh 4c82dbf7e1d8c5ddd70e40b9665cfeee
Description: client support for WPA and WPA2 (IEEE 802.11i)
 wpa-supplicant is a userspace daemon handling connection and authentication
 in wireless and wired networks, primarily secured with the WPA/WPA2/WPA3
 protocols. This software provides key negotiation with the access point
 (WPA Authenticator), and controls association with IEEE 802.11i networks.
Homepage: https://w1.fi/wpa_supplicant/
wlanpi@wlanpi:~/source$ cat /etc/apt/sources.list
deb http://deb.debian.org/debian buster main contrib non-free
#deb-src http://deb.debian.org/debian buster main contrib non-free

deb http://deb.debian.org/debian buster-updates main contrib non-free
#deb-src http://deb.debian.org/debian buster-updates main contrib non-free

deb http://deb.debian.org/debian buster-backports main contrib non-free
#deb-src http://deb.debian.org/debian buster-backports main contrib non-free

deb http://security.debian.org/ buster/updates main contrib non-free
#deb-src http://security.debian.org/ buster/updates main contrib non-free

deb http://deb.debian.org/debian bullseye main contrib non-free
#deb-src http://deb.debian.org/debian bullseye main contrib non-free

deb http://deb.debian.org/debian-security/ bullseye-security main contrib non-free
#deb-src http://deb.debian.org/debian-security/ bullseye-security main contrib non-free

deb http://deb.debian.org/debian bullseye-updates main contrib non-free
#deb-src http://deb.debian.org/debian bullseye-updates main contrib non-free

deb http://deb.debian.org/debian bullseye-backports main contrib non-free
#deb-src http://deb.debian.org/debian bullseye-backports main contrib non-free
                                                                                       
wlanpi@wlanpi:~/source$ cat /etc/apt/sources.list.d/*
deb http://apt.armbian.com buster main buster-utils buster-desktop
deb [trusted=yes] https://apt.fury.io/dfinimundi /
wlanpi@wlanpi:~/source$ 
```

Because I have bullseye and buster repos, I had the choice of which package version to install so I specifically chose 2.9 which wasn't available until I added the bullseye repo.  Here's the wpa_supplicant.conf I was working with:

```
network={
   ssid="ft-test"
   psk=<mypsk>
   key_mgmt=FT-PSK 
   priority=1
}
```
Here's my test network:

![image](https://user-images.githubusercontent.com/52576449/149814452-1b11c55a-9f06-4279-aa99-3fa69d53a683.png)

**Reply by dwrice0** [reply]
@danielmundi your questions made me doubt my original testing methodology.  So, I went back and re-tested and you are correct 802.11R is supported in the packaged 2.9 version out of the box.  I think I may have had underlying version 2.8 wpa_supplicant processes running during my upgrade to 2.9 in my previous testing.  This time I tested the following steps:

1. Fresh wlanpi 2.x image on SD
2. Could not associate to FT only BSS.
3. Added bullseye repo and did apt-get update
4. Installed wpasupplicant package 2.9 and rebooted the wlanpi since I had to join a nonFT BSS to hit the repos.
5. After reboot, it successfully associated to FT only BSS.

**Reply by danielmundi** [reply]
Great to hear that, @dwrice0 !
Do let us know if there are any other features you feel are missing.
